test/simulation/tests/test_node.py

```diff
@@ -0,0 +1,298 @@
+    assert len(nodes[0].list_topic_peer()["listeningShard"]) == 0
+
+
+def test_get_subscribed_shard(unchanged_node):
+    shards = unchanged_node.get_subscribed_shard()
+    assert shards == []
+
+
+def test_subscribe_shard(nodes):
+    # test: wrong type, should be list of int
+    with pytest.raises(TypeError):
+        nodes[0].subscribe_shard(0)
+    # test: normally subscribe
+    nodes[0].subscribe_shard([0])
+    assert nodes[0].get_subscribed_shard() == [0]
+    assert nodes[1].list_shard_peer([0])["0"] == []
+    # test: after adding peers, the node should know the shards its peer subscribes
+    nodes[1].add_peer(nodes[0])
+    time.sleep(0.2)
+    assert nodes[0].peer_id in nodes[1].list_shard_peer([0])["0"]
+    # test: subscribe "0" twice, and shards "1", "2"
+    nodes[0].subscribe_shard([0, 1, 2])
+    assert nodes[0].get_subscribed_shard() == [0, 1, 2]
+    # test: ensure there are two logs `LOG_SUBSCRIBE_SHARD_FINISHED` because of `subscribe_shard`
+    nodes[0].wait_for_log(LOG_SUBSCRIBE_SHARD_FINISHED, 1)
+
+
+def test_unsubscribe_shard(nodes):
+    # test: wrong type, should be list of int
+    with pytest.raises(TypeError):
+        nodes[0].unsubscribe_shard(0)
+    # test: unsubscribe before subscribe
+    nodes[0].unsubscribe_shard([0])
+    # test: normally subscribe
+    nodes[0].subscribe_shard([0])
+    nodes[0].add_peer(nodes[1])
+    nodes[0].unsubscribe_shard([0])
+    assert nodes[0].get_subscribed_shard() == []
+    # test: ensure the peer knows its peer unsubscribes
+    assert nodes[1].list_shard_peer([0])["0"] == []
+    # test: ensure there are two logs `LOG_UNSUBSCRIBE_SHARD_FINISHED` because of
+    #       `unsubscribe_shard`
+    nodes[0].wait_for_log(LOG_UNSUBSCRIBE_SHARD_FINISHED, 1)
+
+
+def test_broadcast_collation(nodes):
+    # test: broadcast to a non-subscribed shard
+    with pytest.raises(CLIFailure):
+        nodes[0].broadcast_collation(0, 1, 100, 123)
+    nodes[0].add_peer(nodes[1])
+    nodes[1].add_peer(nodes[2])
+    nodes[0].subscribe_shard([0])
+    nodes[1].subscribe_shard([0])
+    nodes[2].subscribe_shard([0])
+    time.sleep(2)
+    # test: see if all nodes receive the broadcasted collation. Use a bigger size of collation to
+    #       avoid the possible small time difference. E.g. sometimes t2 < t1, which does not make
+    #       sense.
+    nodes[0].broadcast_collation(0, 1, 1000000, 123)
+    t0 = nodes[0].get_log_time(LOG_BROADCAST_COLLATION_FINISHED, 0)
+    t1 = nodes[0].get_log_time(LOG_RECEIVE_MSG, 0)
+    t2 = nodes[1].get_log_time(LOG_RECEIVE_MSG, 0)
+    t3 = nodes[2].get_log_time(LOG_RECEIVE_MSG, 0)
+    assert t0 < t1
+    assert t1 < t2
+    assert t2 < t3
+
+
+def test_bootstrap(nodes):
+    # test: stop before start
+    nodes[0].bootstrap(False)
+    nodes[0].add_peer(nodes[1])
+    nodes[1].add_peer(nodes[2])
+    time.sleep(0.2)
+    # test: nodes[0] bootstrap and get the nodes[2] as peer with nodes[1] as the bootnode
```

Comment:
nitpick: `# test: nodes[0] bootstraps with nodes[1] as the bootnode and therefore connects to nodes[2]`